drivers/stepper/stepper_stm32.c

```diff
@@ -0,0 +1,653 @@
+
+	return stepper_context_wait_for_completion(&data->ctx);
+}
+
+static int stepper_stm32_set_enabled(struct device *dev, uint32_t channel,
+				     bool enabled)
+{
+	const struct stepper_stm32_config *cfg = to_config(dev);
+	struct stepper_stm32_data *data = to_data(dev);
+
+	int32_t index;
+	const struct stepper_stm32_mconfig *mcfg;
+	struct stepper_stm32_motor_data *mdata;
+
+	index = get_motor_index(cfg, channel);
+	if (index < 0) {
+		LOG_ERR("Invalid channel: %d", channel);
+		return -EINVAL;
+	}
+
+	mcfg = &cfg->motors[index];
+	mdata = &data->motors[index];
+
+	data->enabled = enabled;
+
+	if (data->enabled) {
+		if (IS_TIM_BREAK_INSTANCE(cfg->master_timer)) {
+			LL_TIM_EnableAllOutputs(cfg->master_timer);
+		}
+
+		LL_TIM_CC_EnableChannel(cfg->master_timer, mcfg->ll_channel);
+		gpio_pin_set(mdata->enable, mcfg->enable.pin, true);
+	} else {
+		gpio_pin_set(mdata->enable, mcfg->enable.pin, false);
+		LL_TIM_CC_DisableChannel(cfg->master_timer, mcfg->ll_channel);
+	}
+
+	return 0;
+}
+
+static int stepper_stm32_set_speed(struct device *dev, uint32_t channel,
+				   int32_t speed)
+{
+	const struct stepper_stm32_config *cfg = to_config(dev);
+	struct stepper_stm32_data *data = to_data(dev);
+
+	int32_t index;
+	const struct stepper_stm32_mconfig *mcfg;
+	uint32_t psc, arr;
+
+	index = get_motor_index(cfg, channel);
+	if (index < 0) {
+		LOG_ERR("Invalid channel: %d", channel);
+		return -EINVAL;
+	}
+
+	mcfg = &cfg->motors[index];
+
+	/*
+	 * We have that speed (steps/s) corresponds to the actual PWM frequency,
+	 * f_pwm, given by:
+	 *
+	 * speed = f_pwm = f_tim / ((arr + 1) * (psc + 1)).                  (1)
+	 *
+	 * Duty cycle, d, is given by the ratio between the compare register
+	 * (ccr) and the auto-reload value (arr), that is:
+	 *
+	 * d = ccr / (arr + 1).
+	 *
+	 * As we always want a duty of 50%, we need to set ccr to:
+	 *
+	 * ccr = (arr + 1) / 2.
+	 *
+	 * In case we have a 32-bit timer, we can set psc=0 and set arr as:
+	 *
+	 * arr = (f_tim / speed) - 1
+	 *
+	 * For 16-bit timers the above calculation could overflow the arr
+	 * registers, so prescaler is needed. Maximum resolution is achieved
+	 * when using the full scale of arr (up to 65535). In such case,
+	 * the value of psc that will maximize the resolution is then:
+	 *
+	 * psc = ((f_tim / speed) / (arr + 1)) - 1
+	 *     = ((f_tim / speed) / 65536) - 1
+	 *     = ((f_tim / speed) >> 16) - 1.
+	 *
+	 * Note that the waveform can be generated for arr >= 1, so the maximum
+	 * speed we can generate is given by:
+	 *
+	 * speed_max = f_tim / 2.
+	 *
+	 * NOTE: Thanks to @ABOST for the psc calculation suggestion.
+	 */
+
+	speed *= mcfg->microsteps;
+	if (speed > (data->f_tim / 2U)) {
+		LOG_ERR("Speed out of range: %d", speed);
+		return -EINVAL;
+	}
+
+	psc = 0U;
+	arr = (data->f_tim / speed) - 1U;
+
+	if (!IS_TIM_32B_COUNTER_INSTANCE(cfg->master_timer) && arr > UINT16_MAX) {
+		psc = ((data->f_tim / speed) >> 16U) - 1U;
+		arr = UINT16_MAX;
```

Comment:
I understand this way: we first have `f_tim / speed = K = (arr + 1) * (psc + 1)`. `K` is constant. Either arr or psc can be fixed. If I fix arr to its maximum value, `0xFFFF`, then psc is determined by `psc + 1 = K / (0xFFFF + 1) = K >> 16`. The advantage I see on fixing arr to its maximum value is to always have the maximum PWM resolution (not critical here, though). Am I missing something?

Reply:
Your proposal give better dutycycle resolution (due to lower prescaler) but at a cost of inaccuracy on period (potentially very important) .
Lets take the worst example with your method.
 K = 0x1 FFFF
You fixed arr= 0xFFFF.
then psc = K / (0xFFFF +1) -1 = 0
But reverting the computation: with your method psc = 0 and arr = 0xFFFF 
then  (arr + 1) * (psc + 1) = 0x1000
Delta compare to expected K value DeltaK = 0x1 FFFF - 0x1000 = 0xFFFF
This is almost 50% erreur !!!
With your method, maximum inaccuracy on period is equal to 0xFFFF, because (0XFFFF+ 1) * (psc + 1) will take discrete values 0x10000; 0x20000; 0x30000 ... depending psc.
And there is still issue when K smaller than 0x10000 leading to psc =-1

With my method, I get psc =1 and arr = 0xFFFE
then  (arr + 1) * (psc + 1) = 0xFFFF * 2 = 0x1FFFE
Erreur on period is 1 .
With my method, max inaccuracy on period is equal to prescaler (if prescaler value is N, meaning N tick of tim input, then I may miss a maximum of N tick.)

Reply:
Thanks @ABOSTM for the details. In this case period accuracy is the most important factor, so I see the point. I will fix the calculations, thanks again!